src/overlays/actors/ovl_En_Bubble/z_en_bubble.c

```diff
@@ -32,73 +31,399 @@ const ActorInit En_Bubble_InitVars = {
+    d = (a * a) + (b * b) + (c * c);
+    this->unk1F8.x = a / d;
+    this->unk1F8.y = b / d;
+    this->unk1F8.z = c / d;
+}
+
+u32 func_8089F59C(EnBubble* this) {
+    ColliderInfo* info = &this->colliderSphere.elements[0].info;
+
+    info->toucher.dmgFlags = 8;
+    info->toucher.effect = 0;
+    info->toucher.damage = 4;
+    info->toucherFlags = TOUCH_ON;
+    this->actor.velocity.y = 0.0f;
+    return 6;
+}
+
+u32 func_8089F5D0(EnBubble* this) {
+    EnBubble_SetDimensions(this, -1.0f);
+    return 12;
+}
+
+void EnBubble_DamagePlayer(EnBubble* this, PlayState* play) {
+    play->damagePlayer(play, -this->colliderSphere.elements[0].info.toucher.damage);
+    func_800B8E1C(play, &this->actor, 6.0f, this->actor.yawTowardsPlayer, 6.0f);
+}
+
+s32 EnBubble_Explosion(EnBubble* this, PlayState* play) {
+    u32 i;
+    Vec3f effectAccel;
+    Vec3f effectVel;
+    Vec3f effectPos;
+
+    effectAccel = sEffectAccel;
+    Math_SmoothStepToF(&this->expansionWidth, 4.0f, 0.1f, 1000.0f, 0.0f);
+    Math_SmoothStepToF(&this->expansionHeight, 4.0f, 0.1f, 1000.0f, 0.0f);
+    Math_SmoothStepToF(&this->graphicRotSpeed, 54.0f, 0.1f, 1000.0f, 0.0f);
+    Math_SmoothStepToF(&this->graphicEccentricity, 0.2f, 0.1f, 1000.0f, 0.0f);
+    this->actor.shape.yOffset = ((this->expansionHeight + 1.0f) * 16.0f);
+
+    if (DECR(this->explosionCountdown) != 0) {
+        return -1;
+    }
+    effectPos.x = this->actor.world.pos.x;
+    effectPos.y = this->actor.world.pos.y + this->actor.shape.yOffset;
+    effectPos.z = this->actor.world.pos.z;
+    for (i = 0; i < 20; i++) {
+        effectVel.x = (Rand_ZeroOne() - 0.5f) * 7.0f;
+        effectVel.y = Rand_ZeroOne() * 7.0f;
+        effectVel.z = (Rand_ZeroOne() - 0.5f) * 7.0f;
+        EffectSsDtBubble_SpawnCustomColor(play, &effectPos, &effectVel, &effectAccel, &sEffectPrimColor,
+                                          &sEffectEnvColor, Rand_S16Offset(0x64, 0x32), 0x19, 0);
+    }
+    Item_DropCollectibleRandom(play, NULL, &this->actor.world.pos, 0x50);
+    this->actor.flags &= ~ACTOR_FLAG_1;
+    return Rand_S16Offset(90, 60);
+}
+
+u32 func_8089F8BC(EnBubble* this) {
+    if (DECR(this->explosionCountdown)) {
+        return -1;
+    }
+    return func_8089F5D0(this);
+}
+
+s32 func_8089F908(EnBubble* this) {
+    this->expansionWidth += 1.0f / 12.0f;
+    this->expansionHeight += 1.0f / 12.0f;
+
+    if (DECR(this->explosionCountdown) != 0) {
+        return false;
+    }
+    return true;
+}
+
+void EnBubble_Vec3fNormalizedReflect(Vec3f* vec1, Vec3f* vec2, Vec3f* ret) {
+    f32 norm;
+
+    func_80179F64(vec1, vec2, ret);
+    norm = sqrtf((ret->x * ret->x) + (ret->y * ret->y) + (ret->z * ret->z));
+    if (norm != 0.0f) {
+        ret->x /= norm;
+        ret->y /= norm;
+        ret->z /= norm;
+    } else {
+        ret->x = ret->y = ret->z = 0.0f;
+    }
+}
+
+void EnBubble_Vec3fNormalize(Vec3f* vec) {
+    f32 norm = sqrtf((vec->x * vec->x) + (vec->y * vec->y) + (vec->z * vec->z));
```

Comment:
SQ macro or SQXYZ maybe?
```suggestion
    f32 norm = SQXYZ(vec);
```

Reply:
I don't think so, I could be interpreting it wrong, but to my understanding what is happening here is "norm" on this line is taking the magnitude of the vector, not just squaring the components.

Reply:
You'd have to do `sqrtf(SQXYZ(*vec))`, which isn't completely awful but still a bit nasty, and might make a mess of the codegen. You could replace the multiplications by `SQ`, though

Reply:
I've messed around with this previously, and almost always using the dereference operator with `SQXYZ` breaks codegen. Going the other way does work. i.e. redefining `SQXYZ` using `->` and then either `sqrtf(SQXYZ(vec))` or `sqrtf(SQXYZ(&vec))` works in general, but I've avoided proposing that as I don't really thing that's an improvement.

Reply:
Just use the `SQ`s for now, then.